sql/hive/src/main/scala/org/apache/spark/sql/hive/SQLBuilder.scala

```diff
@@ -0,0 +1,236 @@
+           |
+           |# Original logical plan:
+           |${logicalPlan.treeString}
+           |# Canonicalized logical plan:
+           |${canonicalizedPlan.treeString}
+           |# Built SQL query string:
+           |${maybeSQL.get}
+         """.stripMargin)
+    } else {
+      logDebug(
+        s"""Failed to build SQL query string from given logical plan:
+           |
+           |# Original logical plan:
+           |${logicalPlan.treeString}
+           |# Canonicalized logical plan:
+           |${canonicalizedPlan.treeString}
+         """.stripMargin)
+    }
+
+    maybeSQL
+  }
+
+  private def projectToSQL(
+      projectList: Seq[NamedExpression],
+      child: LogicalPlan,
+      isDistinct: Boolean): Option[String] = {
+    for {
+      listSQL <- sequenceOption(projectList.map(_.sql))
+      childSQL <- toSQL(child)
+      from = child match {
+        case OneRowRelation => ""
+        case _ => " FROM "
+      }
+      distinct = if (isDistinct) " DISTINCT" else ""
+    } yield s"SELECT$distinct ${listSQL.mkString(", ")}$from$childSQL"
+  }
+
+  private def aggregateToSQL(
+      groupingExprs: Seq[Expression],
+      aggExprs: Seq[Expression],
+      child: LogicalPlan): Option[String] = {
+    for {
+      aggSQL <- sequenceOption(aggExprs.map(_.sql))
+      groupingSQL <- sequenceOption(groupingExprs.map(_.sql))
+      maybeGroupBy = if (groupingSQL.isEmpty) "" else " GROUP BY "
+      maybeFrom = child match {
+        case OneRowRelation => ""
+        case _ => " FROM "
+      }
+      childSQL <- toSQL(child).map(maybeFrom + _)
+    } yield {
+      s"SELECT ${aggSQL.mkString(", ")}$childSQL$maybeGroupBy${groupingSQL.mkString(", ")}"
+    }
+  }
+
+  private def toSQL(node: LogicalPlan): Option[String] = node match {
+    case Distinct(Project(list, child)) =>
+      projectToSQL(list, child, isDistinct = true)
+
+    case Project(list, child) =>
+      projectToSQL(list, child, isDistinct = false)
+
+    case Aggregate(groupingExprs, aggExprs, Aggregate(_, _, Expand(_, _, child))) =>
+      aggregateToSQL(groupingExprs, aggExprs, child)
+
+    case Aggregate(groupingExprs, aggExprs, child) =>
+      aggregateToSQL(groupingExprs, aggExprs, child)
+
+    case Limit(limit, child) =>
+      for {
+        limitSQL <- limit.sql
+        childSQL <- toSQL(child)
+      } yield s"$childSQL LIMIT $limitSQL"
+
+    case Filter(condition, child) =>
+      for {
+        conditionSQL <- condition.sql
+        childSQL <- toSQL(child)
+        whereOrHaving = child match {
+          case _: Aggregate => "HAVING"
+          case _ => "WHERE"
+        }
+      } yield s"$childSQL $whereOrHaving $conditionSQL"
+
+    case Union(left, right) =>
+      for {
+        leftSQL <- toSQL(left)
+        rightSQL <- toSQL(right)
+      } yield s"$leftSQL UNION ALL $rightSQL"
+
+    // ParquetRelation converted from Hive metastore table
+    case Subquery(alias, LogicalRelation(r: ParquetRelation, _)) =>
+      Some(s"`$alias`")
```

Comment:
(Here I meant to use `ParquetRelation.metastoreTableName`, which includes database name.)